File: q12-implement-queue-by-stacks/q12-sol.js
// LeetCode 232.
// https://leetcode.com/problems/implement-queue-using-stacks/

class MyQueue {
    constructor() {
        this.stack1 = [];
        this.stack2 = [];
    }
};

// push(1)
// push(2)
// push(3)
// pop()
// pop()
// push(4)
// pop()
// Q  [2,3]
// S1 []
// S2 [3,2]

// pop() => if s2 is empty, clear out the s1 items into s2

/** 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) { // O(1)
    this.stack1.push(x); 
};

/**
 * @return {number}
 */
MyQueue.prototype.pop = function() {  // O(n)
    if (this.stack2.length === 0) {
        while (this.stack1.length) {
            this.stack2.push(this.stack1.pop());
        }
    }
    return this.stack2.pop();
};

/**
 * @return {number}
 */
MyQueue.prototype.peek = function() {  // O(1)
    if (this.stack2.length === 0) return this.stack1[0];
    else return this.stack2[this.stack2.length - 1];
};

/**
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {   // O(1)
    return (this.stack1.length === 0 && this.stack2.length === 0);
};

/** 
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

File: ztm-ds-algo/4-stacks-and-queues/q12-implement-queue-by-stacks/q12-sol.py
# // LeetCode 232.
# // https://leetcode.com/problems/implement-queue-using-stacks/

# MyQueue:   back  [3, 8, 1]  front

# stack1:    front [  ] back
# stack2:    back  [  ] front

class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, x: int) -> None:
        self.stack1.append(x)

    def pop(self) -> int:
        if len(self.stack2) == 0:
            while len(self.stack1) > 0:
                self.stack2.append(self.stack1.pop())
        return self.stack2.pop()  

    def peek(self) -> int:
        if len(self.stack2) > 0:
            return self.stack2[-1]
        if len(self.stack1) > 0:
            return self.stack1[0]
        return None

    def empty(self) -> bool:
        return len(self.stack1) == 0 and len(self.stack2) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()